core/src/main/java/org/infinispan/statetransfer/OutboundTransferTask.java

```diff
@@ -162,119 +148,155 @@ public int getTopologyId() {
+            new AtomicReference<>(Collections.emptyList());
+         subscription =
+            Flowable.concat(publishDataContainerEntries(), publishStoreEntries())
+                    .buffer(chunkSize)
+                    .concatMapCompletable(
+                       // Send the previous batch, not the current one
+                       // This allows us to mark all the segments as finished in the same RPC with the last batch
+                       entries -> {
+                          List<InternalCacheEntry<Object, Object>> previousBatch = batchRef.getAndSet(entries);
+                          if (previousBatch.isEmpty())
+                             return Completable.complete();
+
+                          return completionStageToCompletable().apply(sendEntries(previousBatch, false));
+                       })
+                    .subscribe(() -> {
+                       // Send the remaining entries and mark all the segments as finished
+                       List<InternalCacheEntry<Object, Object>> previousBatch = batchRef.get();
+                       sendEntries(previousBatch, true)
+                          .whenComplete((ignored, throwable) -> {
+                             if (throwable != null) {
+                                logError(throwable);
+                             }
+                             taskFuture.complete(null);
+                          });
+                    }, throwable -> {
+                       logError(throwable);
+                       taskFuture.complete(null);
+                    });
+         if (cancelled) {
+            // In case cancel() was called before we set the subscription field
+            subscription.dispose();
+         }
+         return taskFuture;
+      } catch (Throwable t) {
+         logError(t);
+         taskFuture.complete(null);
+      }
+      return null;
+   }
 
-         AdvancedCacheLoader<Object, Object> stProvider = persistenceManager.getStateTransferProvider();
-         if (stProvider != null) {
-            try {
-               Flowable.fromPublisher(stProvider.entryPublisher(k -> !dataContainer.containsKey(k), true, true))
-                     .blockingForEach(me -> {
-                        int segmentId = keyPartitioner.getSegment(me.getKey());
-                        if (segments.contains(segmentId)) {
-                           try {
-                              InternalCacheEntry entry = mapEntryFromStore.apply(me, entryFactory);
-                              if (entry != null) {
-                                 sendEntry(entry, segmentId);
-                              }
-                           } catch (CacheException e) {
-                              log.failedLoadingValueFromCacheStore(me.getKey(), e);
+   private Flowable<InternalCacheEntry<Object, Object>> publishDataContainerEntries() {
+      return Flowable.fromIterable(() -> dataContainer.iterator(segments))
+                     .flatMapMaybe(ice -> {
```

Comment:
I wonder if is probably worth it to be smart and have a static variable that holds `OutboundTransferTask::defaultMapEntryFromDataContainer` and detect if the passed in `mapEntryFromDataContainer` is the same and just do a `filter` and `map` operation instead of a `flatMapMaybe`. We can avoid N allocations of Maybe then.

Reply:
I'm not even sure we need the filtering, I can't tell if a scattered cache entry is allowed to not have a version, but IMO the allocation is not big enough a deal to add a special case.